python/ray/data/_internal/execution/bundle_queue/base.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import annotations
+
+import abc
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    List,
+    Optional,
+    Protocol,
+    Tuple,
+    runtime_checkable,
+)
+
+if TYPE_CHECKING:
+    from ray.data._internal.execution.interfaces import RefBundle
+
+
+class _QueueMetricRecorderMixin:
+    """Mixin for recording stats about a queue. Subclasses
+    may choose to use the _on_dequeue and _on_enqueue methods to
+    track num_blocks, nbytes, etc... If not, they should override
+    those methods."""
+
+    def __init__(self):
+        self._nbytes: int = 0
+        self._num_blocks: int = 0
+        self._num_bundles: int = 0
+        self._num_rows: int = 0
+
+    def _on_enqueue(self, bundle: RefBundle):
+        self._nbytes += bundle.size_bytes()
+        self._num_blocks += len(bundle.block_refs)
+        self._num_bundles += 1
+        self._num_rows += bundle.num_rows() or 0
+
+    def _on_dequeue(self, bundle: RefBundle):
+        self._nbytes -= bundle.size_bytes()
+        self._num_blocks -= len(bundle.block_refs)
+        self._num_bundles -= 1
+        # Fix this for None count
+        self._num_rows -= bundle.num_rows() or 0
+
+    def estimate_size_bytes(self) -> int:
+        """Return the estimated size in bytes of all bundles."""
+        return self._nbytes
+
+    def num_blocks(self) -> int:
+        """Return the total # of blocks across all bundles."""
+        return self._num_blocks
+
+    def num_rows(self) -> int:
+        """Return the total # of rows across all bundles."""
+        return self._num_rows
+
+    def _reset_metrics(self):
+        self._num_rows = 0
+        self._num_blocks = 0
+        self._num_bundles = 0
+        self._nbytes = 0
+
+    def __len__(self) -> int:
+        """Return the total # bundles."""
+        return self._num_bundles
+
+
+class BaseBundleQueue(_QueueMetricRecorderMixin):
```

Comment:
So the main benefit is that base class should take care of appropriately emitting all of the metrics. 

For that 

 - Make `add` a "final" method & emit stats in it
 - Expose `_add_inner` method that subclasses should override
 - Same for dequeuing